### PR TITLE
Added accurate_alpha parameter to fix blending

### LINKED ISSUE #31124: Image::blend_rect might not blend as expected
Godot 3.1.1

**tl;dr**: `Image::blend_rect()` may cause transparent regions to darken, contrary to what you would get with a drawing app.

Explanation:

When trying to work on a painting app in which you can paint on transparent images using shaders and a transparent viewport, I got to deal with alpha blending. I tried blending a diagonal white-to-transparent-white gradient with the Godot icon, and quickly stumbled on the following problem:

Expected (Paint.NET):
image

Obtained (Godot):
image

Notice the darkened corner. From a pure math standpoint, it's just the result of averaging colors from the "invisible" pixels of the destination image, ending up darkening the gradient. But it may not be the expected result, at least to me it wasn't.
So I went bulldozer mode and blended myself in shaders, coming up with this:
```
vec4 blend_alpha(vec4 a, vec4 b) {
	vec4 res;
	res.rgb = b.a * b.rgb + a.a * a.rgb * (1.0 - b.a);
	res.a = b.a + a.a * (1.0 - b.a);
	res.rgb /= res.a; // THIS
	return res;
}
```
That last division did the trick.

But then, looking at `Image::blend_rect`, I noticed it does the same thing, *except the division*. So I tested it, and it has the same issue.

So is this really expected or should it be patched?

### PR DESCRIPTION
**Pull request by azagaya**
Fixes #31124

### CODE REVIEW

**Comment by azagaya** [comment]
Oh i forgot clang_format again

File: core/image.h [hunk 1]
```diff
@@ -345,8 +345,8 @@ class Image : public Resource {
 
 	void blit_rect(const Ref<Image> &p_src, const Rect2 &p_src_rect, const Point2 &p_dest);
 	void blit_rect_mask(const Ref<Image> &p_src, const Ref<Image> &p_mask, const Rect2 &p_src_rect, const Point2 &p_dest);
-	void blend_rect(const Ref<Image> &p_src, const Rect2 &p_src_rect, const Point2 &p_dest);
-	void blend_rect_mask(const Ref<Image> &p_src, const Ref<Image> &p_mask, const Rect2 &p_src_rect, const Point2 &p_dest);
+	void blend_rect(const Ref<Image> &p_src, const Rect2 &p_src_rect, const Point2 &p_dest, bool accurate_alpha = false);
+	void blend_rect_mask(const Ref<Image> &p_src, const Ref<Image> &p_mask, const Rect2 &p_src_rect, const Point2 &p_dest, bool accurate_alpha = false);
```
Comment (akien-mga) [inline core/image.h:349]: Parameters in C++ prototypes are prefixed with `p_`, so this should be `p_accurate_alpha`.

**Comment by akien-mga** [comment]
I'm not sure that the parameter's name properly conveys what it involves, but otherwise the code looks OK.

Please also add the relevant documentation to `doc/classes/Image.xml`.

**Comment by azagaya** [comment]
> I'm not sure that the parameter's name properly conveys what it involves
`p_accurate_blending` would be better?

**Comment by akien-mga** [comment]
> p_accurate_blending would be better?

Well, why would anyone not want accurate blending in a function named "blend"? :)

**Comment by azagaya** [comment]
> > p_accurate_blending would be better?
> 
> Well, why would anyone want inaccurate blending in a function named "blend"? :)

According to the issue discussing this, this is what i understood. Usually in a game, you blend against a fully opaque background. In those cases, the result without the fix is the same as if it had the fix, because `dc.a` equals 1. So, in those cases, the fix would be an unnecessary loss of performance for each pixel blended.
However, you may want to export some blended images from your game, without fully opaque transparency (like the project uploaded to test the issue). In those cases, not dividing by `dc.a` will give you darkened areas in regions not fully opaque. So, in those cases, the blending is not being accurate.
I initially proposed something like `fix_alpha`, or maybe `fix_alpha_darkening`.
What do you think?

**Comment by bojidar-bg** [comment]
Not sure if it is applicable for this case, but maybe `p_premultipled_alpha` or `p_straight_alpha` might do?

**Comment by azagaya** [comment]
> Not sure if it is applicable for this case, but maybe `p_premultipled_alpha` or `p_straight_alpha` might do?

Well, i think the output without the fix could be thought as the value in "premultiplied alpha". And as the fix divides by alpha, it could be thought that the fix is converting premultiplied to straight. So perhaps the parameter could be called "p_straight_alpha" as you suggest.

**Comment by azagaya** [comment]
Should i rename the parameter to any of those options?

**Comment by azagaya** [comment]
Should i use `Color.blend()` as suggested in the issue? It seems to work fine with that, and that function is already implemented.

**Comment by OverloadedOrama** [comment]
Is there any chance this will be merged before 3.2.2? It would be really helpful for Pixelorama.

**Comment by akien-mga** [comment]
Superseded by #39200.

**Pull request closed without merging**